app/src/main/java/org/akvo/flow/activity/AppUpdateActivity.java

```diff
@@ -175,26 +206,27 @@ protected void onProgressUpdate(Integer... progress) {
             }
 
             Log.d(TAG, "onProgressUpdate() - APK update: " + percentComplete + "%");
-            mProgress.setProgress(percentComplete);
+            updateProgress(percentComplete);
+        }
+
+        private void updateProgress(int percentComplete) {
+            AppUpdateActivity appUpdateActivity = activityWeakReference.get();
+            if (appUpdateActivity != null) {
+                appUpdateActivity.updateProgress(percentComplete);
+            }
         }
 
         @Override
         protected void onPostExecute(String filename) {
-            if (TextUtils.isEmpty(filename)) {
-                Toast.makeText(AppUpdateActivity.this, R.string.apk_upgrade_error, Toast.LENGTH_SHORT).show();
-                mInstallBtn.setText(R.string.retry);
-                mInstallBtn.setEnabled(true);
-                return;
+            if (activityWeakReference.get() != null) {
```

Comment:
I'm no expert in weak references, but isn't it theoretically possible that the activity could be garbage collected between these two calls to get() ? To close this loophole you need to assign the result of the (first) get() to a variable, which you then test **and** use.